doc/AEPs/aep-0002.md

```diff
@@ -0,0 +1,200 @@
+that should be of no concern to users of Alire, just like the private part of
+the specification holds no interest for users of an Ada package.  This is
+detailed further in the [Implementation](#implementation) section.
+
+This `[private]` table could be leveraged in the future for other information
+besides that inserted by `alr with`. For example, the lockfile contents could
+be moved there to avoid exposing another file that is not user editable anyway.
+
+Advantages for packagers are:
+
+1. They can format/comment the manifest without their editions being lost.
+1. The whole publish process can be automated (see the
+[examples](#workflow-examples) section).
+
+Advantages for regular users:
+
+1. They can use `alr with` as currently, without manually editing the manifest
+(for now, for platform-independent crates).
+1. They can instruct their users to simply use `alr update`, `alr build`, or
+`alr run` to obtain dependencies, build, or run their projects, respectively.
+This is analogous to `cargo` usage.
+
+Why not `requirements.txt`
+--------------------------
+
+Another possibility would be to go pip's way and have something similar to its
+[requirements.txt](#references) file. This file contains a sequence of commands
+to pass to `pip install` to recreate the build environment for a project.
+
+The analogy for Alire would be to have a similar file with commands for `alr`.
+This has a few drawbacks with no clear advantage over `cargo`'s solution:
+
+1. It does not help packagers, unless we have complete manipulation of the 
+manifest file via `alr` commands.
+    1. This would be a huge undertaking, just to generate a file that is 
+easily maintained by hand.
+    1. We would have in practice a duplication of metadata; one copy in the
+manifest file and another copy in the requirements file, in two different 
+formats.
+    1. Same problems with manifest formatting/comments being lost.
+
+Implementation
+==============
+
+The proposal relies on the following changes:
+
+- `${crate}/alire/crate.toml` (manifest) is under version control and manually 
+edited.
+- A new machine-generated table is added at the end of the manifest. Its
+  contents are opaque to users to all intents and purposes.
+
+At present, dependencies in a release can be provided in the `[general]` or in
+a version-specific table. Similarly, dependencies edited through `alr with`
+will be stored in the `private.depends-on` table and become a new source of
+dependencies for the release.
+
+A consequence is that dependencies added manually (editing the public part of
+the manifest) can not be removed using `alr with`. This is the (minor IMHO)
+toll to pay to keep the `alr with` command. (Low-level processing of the
+manifest could probable work around this anyway, so that could be a
+low-priority enhancement).
+
+Migration plan
+--------------
+
+A problem with the current situation is that packaged releases do not currently
```

Comment:
That means you want to use the manifest in the source package of a release for instance, rather than use the one already in the index?

Reply:
If I understand the context you quote correctly, yes, but this is only a concern for a root crate retrieved with `alr get`.

For the root crate we rely exclusively on the on-disk manifest. So, for `alr get` crates we currently regenerate the local manifest from the index information and from then on is as if the user had provided the manifest. So the idea is to actually use the one they provide at the time of packaging (that should match the one published).

Dependencies OTOH use the index information from which we find and retrieve them; any manifest they include is ignored (and we do not recreate one for them). (Except for dependencies linked from a directory; the local manifest is used because there's no corresponding index information.)

Reply:
> use the one they provide at the time of packaging (that should match the one published)

That is my concern, we don't know if the manifest packaged in a release is going to be the same as the one in the index.

Reply:
This is also on my radar. Thoughts:

- Compare local and remote manifest and warn/err if they differ.
   - We should check this too server-side for new submissions, so the logic could be reused.
- Use the index manifest by default after `alr get`, until a manual edition to the local manifest is done.
   - Overkill if we have checks, simpler if we don't have.

As far as I see, this is only a concern for root crates retrieved via `alr get`. Otherwise the local manifest is the only source of metadata.